firedrake/interpolation.py

```diff
@@ -227,17 +234,49 @@ def _interpolator(V, tensor, expr, subset, arguments, access):
         raise RuntimeError('Shape mismatch: Expression shape %r, FunctionSpace shape %r'
                            % (expr.ufl_shape, V.ufl_element().value_shape()))
 
-    mesh = V.ufl_domain()
-    coords = mesh.coordinates
+    target_mesh = V.ufl_domain()
+    try:
+        trans_mesh = (
+            expr.ufl_domain() is not None  # Constant expr
+            and expr.ufl_domain() != V.mesh()  # Coming from a different domain
+        )
+    except AttributeError:
+        # Have python Expression
+        trans_mesh = False
+
+    if trans_mesh:
+        if target_mesh.geometric_dimension() != expr.ufl_domain().geometric_dimension():
+            raise ValueError("Cannot interpolate onto a mesh of a different geometric dimension")
+        # For trans-mesh interpolation we use a FInAT QuadratureElement as the
+        # (base) target element with runtime point set expressions as their
+        # quadrature rule point set and weights from their dual basis.
+        # NOTE: This setup is useful for thinking about future design - in the
+        # future this `rebuild` function can be absorbed into FInAT as a
+        # transformer that eats an element and gives you an equivalent (which
+        # may or may not be a QuadratureElement) that lets you do run time
+        # tabulation. Alternatively (and this all depends on future design
+        # decision about FInAT how dual evaluation should work) the
+        # to_element's dual basis (which look rather like quadrature rules) can
+        # have their pointset(s) directly replaced with run-time tabulated
+        # equivalent(s) (i.e. finat.point_set.UnknownPointSet(s))
+        rt_var_name = 'rt_X'
+        to_element = rebuild(to_element, expr, rt_var_name)
+        # The source domain is on the expression not the FunctionSpace/Function
+        # we are interpolating into/onto
+        source_coords_coeff = expr.ufl_domain().coordinates
+        source_domain = expr.ufl_domain()
+    else:
+        # The source domain is that of the FunctionSpace/Function we are
+        # interpolating into/onto
+        source_coords_coeff = target_mesh.coordinates
+        source_domain = V.mesh()
```

Comment:
```
source_domain = target_mesh
source_coords_coeff = target_mesh.coordinates
```
Maybe `source_domain` should be renamed `source_mesh`, or can these two object be just called `mesh` and `coords`?

Reply:
Tidied. Though I think keeping a clear distinction between source and target in the naming is helpful. I found myself getting mixed up until I added the prefixes.